---
title: "RNA-seq normalisation"
output: html_notebook
---

```{r knitr-setup, echo=FALSE}
knitr::opts_chunk$set(fig.path = 'figures/',
                      dev = c('png', 'pdf'),
                      dpi = 100)
```

```{r setup, message=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
```

```{r ggplot-setup, echo=FALSE}
theme_set(theme_minimal())

nogrid = function ()
    theme(panel.grid.minor = element_blank(),
          panel.grid.major.x = element_blank())

pie_chart = function ()
    list(coord_polar(theta = 'y', direction = -1),
         theme(axis.text = element_blank(),
               panel.grid = element_blank()))
```

```{r rna-seq-counts}
mean_fragment_length = 300

counts = tibble(Gene = c('A', 'B', 'C', 'D'),
                Length = c(700, 700, 800, 800),
                Control = c(40, 20, 30, 40),
                Treatment = c(20, 20, 30, 40) * 2) %>%
    gather(Library, Count, -Gene, -Length) %>%
    mutate(EffectiveLength = Length - mean_fragment_length + 1)

counts

gene_colors = c(A = 'dodgerblue2', B = 'darkgrey', C = 'darkgrey', D = 'darkgrey')
```

## RNA-seq gene expression by counts

```{r rna-seq-barchart, fig.width=6, fig.height=3}
ggplot(counts, aes(x = Gene, y = Count, fill = Gene)) +
    geom_bar(stat = 'identity') +
    facet_wrap(~ Library) +
    scale_fill_manual(values = gene_colors, guide = FALSE) +
    labs(y = 'RNA molecules') +
    nogrid()
```

## RNA-seq gene expression as fractions

```{r rna-seq-piechart, fig.width=6, fig.height=3}
# FIXME: Direction should be clockwise but `direction = 1` does anticlockwise
# on my computer. But labels are clockwise?!
counts %>%
    group_by(Library) %>%
    mutate(Fraction = Count / sum(Count)) %>%
    mutate(LabelPos = sum(Fraction) - (cumsum(Fraction) - Fraction / 2)) %>%
    ggplot(aes(x = '', y = Fraction, fill = Gene)) +
    geom_bar(stat = 'identity', width = 1, color = 'white') +
    geom_text(aes(y = LabelPos, label = Gene)) +
    pie_chart() +
    facet_wrap(~ Library) +
    scale_fill_manual(values = gene_colors, guide = FALSE) +
    labs(x = '', y = 'Fraction of RNA molecules')
```

---

> **TPM** (transcripts per millions):
> Given a million RNA fragments, how many fragments of a given transcript do we see?

TPMs for a transcript \(i\) are calculated thusly:

\[
\text{TPM}_\color{red}i =
    {\color{dodgerblue}{\frac{x_\color{red}i}{{l_\text{eff}}_\color{red}i}}}
    \cdot
    \frac{1}{\sum_\color{tomato}j \color{dodgerblue}{\frac{x_\color{tomato}j}{{l_\text{eff}}_\color{tomato}j}}}
    \cdot
    \color{darkcyan}{10^6}
\]

Where \(\color{red}i\) is the index of a transcript, and \(\color{tomato}j\) iterates over all (known) transcripts. \(\color{dodgerblue}{\frac{x_k}{{l_\text{eff}}_k}}\) is the *rate of fragment coverage per nucleobase*. Finally, \(\color{darkcyan}{10^6}\) is a scaling factor (= “per millions”).
