---
title: "RNA-seq expression normalisation"
author: "Konrad Rudolph (<a href='https://twitter.com/klmr'>@klmr</a>)"
output:
  revealjs::revealjs_presentation:
    center: yes
    css: presentation.css
    incremental: yes
    theme: league
    transition: none
  html_notebook: default
  html_document: default
---

```{r knitr-setup, echo=FALSE}
presentation_mode = identical(knitr::opts_knit$get('rmarkdown.pandoc.to'), 'revealjs')

knitr::opts_chunk$set(fig.path = 'figures/',
                      fig.width = 3,
                      fig.height = 3,
                      dev = 'png',
                      dpi = 100,
                      echo = ! presentation_mode,
                      results = if (presentation_mode) 'hide' else 'markup')
```

```{r setup, message=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
```

```{r ggplot-setup, echo=FALSE}
theme_set(theme_minimal())

nogrid = function ()
    theme(panel.grid.minor = element_blank(),
          panel.grid.major.x = element_blank())

pie_chart = function ()
    list(coord_polar(theta = 'y', direction = -1),
         theme(axis.text = element_blank(),
               panel.grid = element_blank()))

brackets = function (...) {
    args = list(...)
    grid:::recordGrob(do.call(pBrackets::grid.brackets, args), environment())
}
```

```{r rna-seq-counts}
mean_fragment_length = 300

counts = tibble(Gene = c('A', 'B', 'C', 'D'),
                Length = c(700, 700, 800, 800),
                Control = c(40, 20, 30, 40),
                Treatment = c(20, 20, 30, 40) * 2) %>%
    gather(Library, Count, -Gene, -Length) %>%
    mutate(EffectiveLength = Length - mean_fragment_length + 1)

counts

gene_colors = c(A = 'dodgerblue2', B = 'darkgrey', C = 'darkgrey', D = 'darkgrey')
```

## RNA-seq gene expression by counts

```{r rna-seq-barchart, fig.width=6, fig.height=3}
ggplot(counts, aes(x = Gene, y = Count, fill = Gene)) +
    geom_bar(stat = 'identity') +
    facet_wrap(~ Library) +
    scale_fill_manual(values = gene_colors, guide = FALSE) +
    labs(y = 'Sequenced fragments') +
    nogrid()
```

---

But: we don’t want  
<span class="em">**fragments per library**</span>.

We want  
<span class="em">**RNA molecules per cell**</span>.

## Library size effect

```{r library-size-barchart}
ggplot(counts, aes(x = Library, y = Count, fill = Gene)) +
    geom_bar(stat = 'sum') +
    scale_fill_manual(values = gene_colors) +
    labs(y = 'Sequenced fragments') +
    nogrid() +
    theme(legend.position = 'none')
```

## Sources of variability

* Across libaries
    * Amount of RNA sequenced
    * Technical noise
* Across transcripts
    * … longer transcripts = more fragments
* <span class="em">Biologically meaningful</em>

## RNA-seq gene expression as fractions

```{r rna-seq-piechart, fig.width=6, fig.height=3}
# FIXME: Direction should be clockwise but `direction = 1` does anticlockwise
# on my computer. But labels are clockwise?!
counts %>%
    group_by(Library) %>%
    mutate(Fraction = Count / sum(Count)) %>%
    mutate(LabelPos = sum(Fraction) - (cumsum(Fraction) - Fraction / 2)) %>%
    ggplot(aes(x = '', y = Fraction, fill = Gene)) +
    geom_bar(stat = 'identity', width = 1, color = 'white') +
    geom_text(aes(y = LabelPos, label = Gene)) +
    pie_chart() +
    facet_wrap(~ Library) +
    scale_fill_manual(values = gene_colors, guide = FALSE) +
    labs(x = '', y = 'Fraction of RNA molecules')
```

## Transcript length

```{r transcript-length, fig.width=6, fig.height=2.5}
tlen = 22
flen = 10
gap = 0.3
transcript = tibble(xmin = 1, ymin = 1, xmax = tlen, ymax = 2, type = 'transcript')
fragments = tibble(xmin = seq(1, tlen - flen), ymin = xmin * (1 + gap) + 1,
                   xmax = xmin + flen, ymax = ymin + 1, type = 'fragment')

annotate_equation = function (x, y, label)
    annotate('text', x = x, y = y, label = label, parse = TRUE,
             size = 8, family = 'Times New Roman')

ggplot(bind_rows(transcript, fragments),
       aes(xmin = xmin, ymin = ymin, xmax = xmax, ymax = ymax, fill = type)) +
    geom_rect() +
    coord_cartesian(xlim = c(-1.5, tlen), ylim = c(-3, tail(fragments$ymax, 1))) +
    annotation_custom(brackets(0.135, 0.28, 0.135, 1, h = 0.1, lwd = 2)) +
    annotate_equation(-1.5, 11, 'italic(l)[plain(eff)]') +
    annotation_custom(brackets(0.145, 0.21, 0.95, 0.21, h = -0.1, lwd = 2)) +
    annotate_equation(11.4, -2.9, 'italic(l)') +
    annotation_custom(brackets(0.145, 0.34, 0.526, 0.34, h = 0.1, lwd = 2)) +
    annotate_equation(6.15, 7.5, 'italic(bar(n))') +
    scale_fill_manual(values = c(transcript = 'darkolivegreen4', fragment = 'darkgrey')) +
    theme(panel.grid = element_blank(),
          legend.position = 'none',
          axis.title = element_blank(),
          axis.text = element_blank())
```

\[
\large
l_\text{eff} = l - \bar{n} + 1
\]

---

Different transcript lengths lead to different expected fragment counts

```{r simulate-fragments}
transcripts = IRanges::IRanges(start = c(1, 22), width = c(20, 40))

fstarts = IRanges::resize(transcripts, IRanges::width(transcripts) - flen + 1) %>%
    IRanges::coverage() %>%
    {seq_along(.)[as.integer(.) == 1]}

set.seed(1234)
fragments = IRanges::IRanges(start = sample(fstarts, 9), width = flen)
fragment_data = fragments %>%
    as.data.frame() %>%
    mutate(bin = IRanges::disjointBins(fragments), type = 'read')
```

```{r transcript-length-differences, fig.width=8, fig.height=2}
transcripts %>%
    as.data.frame() %>%
    mutate(bin = 0, type = c('a', 'b')) %>%
    bind_rows(fragment_data) %>%
    mutate(ymin = bin * (1 + gap), ymax = ymin + 1) %>%
    ggplot(aes(xmin = start, ymin = ymin, xmax = end, ymax = ymax, fill = type)) +
    geom_rect() +
    scale_fill_manual(values = c(a = 'darkolivegreen4', b = 'dodgerblue3', read = 'darkgrey')) +
    theme(panel.grid = element_blank(),
          legend.position = 'none',
          axis.title = element_blank(),
          axis.text = element_blank())
```

> **TPM** (transcripts per millions):
> Given a million RNA fragments, how many fragments of a given transcript do we see?

---

\[
\large
\text{TPM}_\color{orchid}i =
    {\color{dodgerblue}{\frac{x_\color{orchid}i}{{l_\text{eff}}_\color{orchid}i}}}
    \cdot
    \frac{1}{\sum_\color{tomato}j \color{dodgerblue}{\frac{x_\color{tomato}j}{{l_\text{eff}}_\color{tomato}j}}}
    \cdot
    \color{darkcyan}{10^6}
\]

* \(\color{orchid}i\): index of a transcript,
* \(\color{tomato}j\) iterates over all (known) transcripts,
* \(\color{dodgerblue}{\frac{x_k}{{l_\text{eff}}_k}}\): rate of fragment coverage per nucleobase,
* \(\color{darkcyan}{10^6}\): scaling factor (= “per millions”).

```{r tpm}
tpm = function (counts, effective_lengths) {
    rate = log(counts) - log(effective_lengths)
    exp(rate - log(sum(exp(rate))) + log(10 ^ 6))
}
```

```{r tpms}
tpms = counts %>%
    group_by(Library) %>%
    mutate(TPM = tpm(Count, EffectiveLength))

tpms
```

## RNA-seq gene expression in TPM

```{r rna-seq-tpm-barchart, fig.width=6, fig.height=3}
ggplot(tpms, aes(x = Gene, y = TPM, fill = Gene)) +
    geom_bar(stat = 'identity') +
    facet_wrap(~ Library) +
    scale_fill_manual(values = gene_colors, guide = FALSE) +
    labs(y = 'TPM') +
    nogrid()
```
